---
title: "R Notebook"
output: html_notebook
params:
  out_prefix: "~/mount/CT/git_test/results/"
  antibody: 'scATAC_P50'
---

```{r libraries}
library(Seurat)
library(Signac)
library(EnsDb.Mmusculus.v79)
library(ggplot2)
library(viridis)
library(GenomicRanges)
set.seed(2020)
```


```{r source_functions}
source("../../scripts/R/func.R")
```



```{r}
counts <- Read10X_h5(paste0(params$out_prefix,"scATAC_P50/data/atac_v1_adult_brain_fresh_5k_filtered_peak_bc_matrix.h5"))
metadata <- read.csv(
  file = paste0(params$out_prefix,"scATAC_P50/data/atac_v1_adult_brain_fresh_5k_singlecell.csv"),
  header = TRUE,
  row.names = 1
)

fragment.path <- paste0(params$out_prefix,"scATAC_P50/data/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz")

brain_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  genome = 'mm10',
  fragments = fragment.path,
  min.cells=1
  )

brain <- CreateSeuratObject(
  counts = brain_assay,
  assay = 'peaks',
  meta.data = metadata
)


```

```{r}
brain <- RunTFIDF(brain)
brain <- FindTopFeatures(brain, min.cutoff = 'q0')
brain <- RunSVD(
  object = brain,
  assay = 'peaks',
  reduction.key = 'LSI_',
  reduction.name = 'lsi'
)

brain <- RunUMAP(object = brain, reduction = 'lsi', dims = 2:30)
brain <- FindNeighbors(object = brain, reduction = 'lsi', dims = 2:30)
brain <- FindClusters(object = brain, verbose = FALSE, algorithm = 3)
DimPlot(object = brain, label = TRUE) + NoLegend()
```


```{r load_gene_annotations}
genebodyandpromoter.coords.flat <- load_ensembl_annot()
brain <- createFeatureAssay(seurat_object = brain,
                            features = genebodyandpromoter.coords.flat,
                            assay_name = "GA")

```

```{r}
markers       <- read.csv2(file=paste0(params$out_prefix,'Sten_RNA/clustering/sten_RNA_markers.csv'),row.names = 1)
assay='GA'



meta.df <- metageneScoreFromMarkers(seurat_object = brain,
                                    markers_df = markers,
                                    assay = assay)
brain <- AddMetaData(brain,metadata = meta.df,col.name = colnames(meta.df))



p.ls <- lapply(as.character(colnames(meta.df)),function(x){
  FeaturePlot(brain,x,max.cutoff = quantile(meta.df[,x],0.9,na.rm=TRUE),min.cutoff = quantile(meta.df[,x],0.05,na.rm=TRUE)) +
    scale_color_viridis() + NoLegend()
  })


```



```{r}
# Manual re-labeling
# brain <- RenameIdents(brain,'4' = "mOL",'8' = 'mOL','12' = 'OPC','5' = "Astrocytes",'10'='Astrocytes')
# DimPlot(brain,label=TRUE) + NoLegend()
```

```{r,fig.width=10,fig.height=10}
brain.RNA              <- readRDS(paste0(params$out_prefix,'Sten_RNA/clustering/01.clustering_20000cells.Rds'))
brain.RNA              <- SetIdent(brain.RNA,cells = colnames(brain.RNA),value = as.factor(gsub(" ","_",brain.RNA$TaxonomyRank4)))

# Reset scATAC cluster names
brain@active.ident <- as.factor(brain$seurat_clusters)

DimPlot(brain.RNA,label=TRUE,repel = TRUE) + NoLegend()
DimPlot(brain,label=TRUE) + NoLegend()
```

```{r,fig.width=14,fig.height=14}
DefaultAssay(brain) <- "GA"
# features = intersect(rownames(brain),rownames(brain.RNA))
brain.RNA <- FindVariableFeatures(object = brain.RNA,nfeatures = 5000)

transfer.anchors <- FindTransferAnchors(
  reference = brain.RNA,
  query = brain,
  reduction = 'cca')# ,k.filter = NA)

predicted.labels <- TransferData(
  anchorset = transfer.anchors,
  refdata = brain.RNA@active.ident,
  weight.reduction = brain[['lsi']],
  dims = 2:30
)

brain <- AddMetaData(object = brain, metadata = predicted.labels)
plot1 <- DimPlot(
  object = brain.RNA,
  group.by = 'TaxonomyRank4',
  label = TRUE,
  repel = TRUE) + NoLegend() + ggtitle('scRNA-seq')

plot2 <- DimPlot(
  object = brain,
  group.by = 'predicted.id',
  label = TRUE,
  repel = TRUE) + NoLegend() + ggtitle('scATAC-seq')

plot1 + plot2
```

```{r}

chrom.sizes <- read.table(url('http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/mm10.chrom.sizes'),sep="\t",stringsAsFactors = FALSE)
chrom.sizes <- chrom.sizes[1:21,]
fragments <- rtracklayer::import(fragment.path,format = 'bed')

brain@active.ident <- as.factor(brain$predicted.id)

dir.create(paste0(params$out_prefix,params$antibody,"/clustering/bigwig/"),recursive = TRUE)
       
to.export.bw <- list("Oligodendrocytes","Astrocytes","Oligodendrocyte_precursor_cells")
    
lapply(to.export.bw,function(x){
  exportBW(brain,x,fragments, paste0(params$out_prefix,params$antibody,"/clustering/bigwig/cluster_",x,".bw"))
})

```

```{r}
cluster.annotations           <- as.data.frame(brain@active.ident[brain@active.ident %in% c('Oligodendrocytes','Astrocytes','Oligodendrocyte_precursor_cells','Vascular_and_leptomeningeal_cells')])

cluster.annotations$barcode   <- rownames(cluster.annotations)
rownames(cluster.annotations) <- NULL
colnames(cluster.annotations) <- c('#cluster', '#barcode')

write.csv(x = cluster.annotations, file=paste0(params$out_prefix,params$antibody,'/clustering/bam_per_cluster/cluster_barcode_table.csv'),row.names = FALSE,quote = FALSE,)

```


```{r}
saveRDS(object = brain,file = paste0(params$out_prefix,params$antibody,"/clustering/clustering_scATAC.Rds"))
```



















