---
title: "R Notebook"
output:
  html_document:
    df_print: paged
params:
  out_prefix: "~/mount/CT/git_test/results/"
---

```{r}
library(Seurat)
library(Signac)
library(ggplot2)
#library(EnsDb.Mmusculus.v79)
#library(ensembldb)
#library(GenomicRanges)
#library(dplyr)
#library(gridExtra)
set.seed(2020)
```


```{r Create_marques_seurat_object}
brain.RNA.matrix      <- read.table(gzfile(paste0(params$out_prefix,'marques_RNA/data/GSE75330_Marques_et_al_mol_counts2.tab.gz')),stringsAsFactors = FALSE)
brain.RNA.annotations <- readRDS(file =paste0(params$out_prefix,'marques_RNA/data/Marques2016annotation.rds'))

colnames(brain.RNA.matrix) <- gsub("-","_",sub('-','',gsub("C1-","",brain.RNA.matrix[1,])))
brain.RNA.matrix           <- brain.RNA.matrix[-1,]
rownames(brain.RNA.matrix) <- brain.RNA.matrix[,1]
brain.RNA.matrix           <- brain.RNA.matrix[,-1]

brain.RNA <- CreateSeuratObject(counts = brain.RNA.matrix,meta.data = brain.RNA.annotations,assay = 'RNA')
```

```{r seurat_dimreduce}
all.genes <- rownames(brain.RNA)

brain.RNA <- NormalizeData(brain.RNA, normalization.method = "LogNormalize", scale.factor = 10000)
brain.RNA <- FindVariableFeatures(brain.RNA, selection.method = "vst", nfeatures = 2000)
brain.RNA <- ScaleData(brain.RNA, features = all.genes)
brain.RNA <- RunPCA(brain.RNA, features = VariableFeatures(object = brain.RNA))

brain.RNA <- RunUMAP(brain.RNA, dims = 1:20)
DimPlot(brain.RNA, reduction = "umap",group.by = 'cell_class')

```

```{r add_metadata}
brain.RNA$cell_type <- brain.RNA$cell_class
brain.RNA$antibody <- 'RNA-seq'
```

```{r find_markers, fig.width=12,fig.height=12}
brain.RNA@active.ident <- brain.RNA$cell_type
markers                <- FindAllMarkers(brain.RNA)
markers.pos            <- markers[markers$p_val < 0.05 & markers$avg_logFC > 0.5,]


# Merge NFOL and MFOL identities
new_ids <- as.character(brain.RNA$cell_class)
new_ids[new_ids == 'NFOL2'] <- 'NFOL'
new_ids[new_ids == 'NFOL1'] <- 'NFOL'
new_ids[new_ids == 'MFOL2'] <- 'MFOL'
new_ids[new_ids == 'MFOL1'] <- 'MFOL'

brain.RNA <- SetIdent(brain.RNA,cells = names(new_ids),value = new_ids)
DimPlot(brain.RNA)

markers.merged <- FindAllMarkers(brain.RNA)
```

```{r export}
# Export
write.csv2(x = markers, file = paste0(params$out_prefix,"marques_RNA/clustering/markers.csv"))
write.csv2(x =  markers.merged,file = paste0(params$out_prefix,"marques_RNA/clustering/markers_merged_clusters.csv"))
saveRDS(object = brain.RNA,file = paste0(params$out_prefix,"marques_RNA/clustering/01.clustering.Rds"))
```
















