---
title: "R Notebook"
output:
  html_document:
    df_print: paged
params:
  config: '../config/config.yaml'
  #out_prefix: "/Users/mareba/mount/CT/git_test/results/" #'results/'
  out_prefix: "/data/proj/GCB_MB/CT/git_test/results/" #'results/'
---

# UMAP

```{r}
library(Seurat)
library(ggplot2)
library(ggthemes)
library(viridis)
library(Signac)
library(reshape2)
library(scales)
library(reticulate)
library(ggrepel)



brain <- readRDS(paste0(params$out_prefix,"H3K27me3_cell_lines/clustering/01.clustering.Rds"))

DimPlot(brain) + NoLegend()
FeaturePlot(brain,'logUMI') + scale_colour_viridis_c()

DimPlot(brain) + scale_color_manual(values =  c("mESC_1"='deepskyblue1',"mESC_2"='royalblue4',"3T3"='seagreen',"Oli-neu_1"='violet',"Oli-neu_2"='red3')) + NoAxes() + NoLegend()

dir.create(paste0(params$out_prefix,"Figures/revision/"),recursive = TRUE)
ggsave(paste0(params$out_prefix,"Figures/revision/Cell_lines_UMAP.png"),width=6,height=6)
```

```{r}
ggplot(brain@meta.data) + geom_boxplot(aes(x=cell_type,y=logUMI))
```


######### FriP analysis ploting

```{r}

prefix   <- paste0(params$out_prefix,'other_datasets/frip_analysis/')
files    <- c('scCT/H3K27me3_N1/','scCT/H3K27me3_N2/','scCT/H3K27me3_N3/','scCT/H3K27me3_N4/','kaya_okur/K562_H3K4me2_iCell8/','kaya_okur/K562_H3K27me3_iCell8/','kaya_okur/H1_H3K27me3_iCell8/','scCT/H3K27me3_cell_lines_1/','scCT/H3K27me3_cell_lines_2/')

group    <- c(rep("scCT_brain",4),rep("Kaya-Okur",3),rep('scCT_cell_lines',2))
sample   <- gsub("scCT|kaya_okur|/","",files)
antibody <- c(rep('H3K27me3',4),'H3K4me2',rep('H3K27me3',4))

all.sufix   <- 'all_fragments.txt'
peaks.sufix <- 'peak_fragments.txt'

barcodes_scCT       <- read.csv(file=paste0(params$out_prefix,'H3K27me3/clustering/bam_per_cluster/cluster_barcode_table.csv'),stringsAsFactors=FALSE)
barcodes_scCT_lines <- read.csv(file=paste0(params$out_prefix,'H3K27me3_cell_lines/clustering/bam_per_cluster/cluster_barcode_table.csv'),stringsAsFactors=FALSE)


files.df        <- data.frame(path=paste(prefix,files,sep=''))
files.df$all    <- paste(files.df$path,all.sufix,sep='')
files.df$peak   <- paste(files.df$path,peaks.sufix,sep='')
files.df$group  <- group
files.df$sample <- sample
files.df$antibody <- antibody


f.ls <- apply(files.df,1,function(x){
  x.all  <- read.table(file=x['all'],stringsAsFactors=FALSE)
  x.peak <- read.table(file=x['peak'],stringsAsFactors=FALSE)
  x.df   <- merge(x.all,x.peak,by='V2')
  colnames(x.df) <- c("barcode","all","peak")
  
  x.df$ratio <- x.df$peak / x.df$all
  
  x.df$group <- x['group']
  x.df$sample <- x['sample']
  x.df$antibody <- x['antibody']
  
  if(x['group'] == 'scCT_brain'){
    x.df <- x.df[x.df$barcode %in% as.character(barcodes_scCT[,2]),]
  }
  if(x['group'] == 'scCT_cell_lines'){
    x.df <- x.df[x.df$barcode %in% as.character(barcodes_scCT_lines[,2]),]
  }
  
  x.df
})




f.merge <- do.call('rbind',f.ls)
f.merge$sample <- factor(f.merge$sample,levels=unique(f.merge$sample)[c(1:4,8,9,5,6,7)])
f.merge$group  <- factor(f.merge$group,levels=unique(f.merge$group)[c(1,3,2)])
```

```{r}
sra   <- c("SRR7536860","SRR7536861","SRR7536862")
#names <- 

grosselin.frip <- lapply(sra,function(x){
  read.table(file=paste0(params$out_prefix,'other_datasets/grosselin_scChIP/',x,'/frip/',x,'_cell_statistics.txt'),stringsAsFactors=FALSE)
})

grosselin.barcodes <- lapply(sra,function(x){
  as.character(read.table(file=paste0(params$out_prefix,'other_datasets/grosselin_scChIP/',x,'/barcodes/counts/',x,'_flagged_rmPCR_RT_rmDup_counts_50000.tsv'),stringsAsFactors=FALSE,nrows = 1)[1,-1])
})


grosselin.frip.filter<- lapply(1:length(grosselin.frip),function(x){
  d <- grosselin.frip[[x]]
  d$V1 <- gsub("XB:Z:","",d$V1)
  d <- d[d$V1 %in% grosselin.barcodes[[x]],]
  d
})


grosselin.frip.filter <-  lapply(1:length(grosselin.frip.filter),function(x){
  d <- grosselin.frip.filter[[x]]
  d$sample <- paste0("Grosselin_",x)
  d
})

lapply(grosselin.frip.filter,function(x){median(x$V4)})

grosselin.to.plot <- do.call('rbind',grosselin.frip.filter)
colnames(grosselin.to.plot) <- c("barcode","peak","all","ratio","sample")

grosselin.to.plot$group    <- "scChIP-seq"
grosselin.to.plot$antibody <- "H3K27me3"

f.merge <- rbind(f.merge,grosselin.to.plot)
```



```{r,fig}

p1 <- ggplot(data=f.merge,aes(y=ratio,x=sample,fill=group)) + 
  geom_violin(trim = TRUE,scale = 'width')  + 
  theme_few() + 
  theme(axis.text.x = element_text(angle = 60,hjust=1))+ xlab("") +
  scale_fill_discrete(name="") + 
  stat_summary(fun.data = "mean_sdl",geom='pointrange',fun.args = (mult=1),color='black',position = position_dodge(0.9)) + 
  ylab("Fraction of fragments in peaks") + 
  theme(text=element_text(size=18))

p2 <- ggplot(data=f.merge,aes(y=log10(all),x=sample,fill=group)) + 
  geom_violin(outlier.shape = NA,scale = 'width') + 
  theme_few() + 
  theme(axis.text.x = element_text(angle = 60,hjust=1)) + 
  xlab("") + 
  scale_fill_discrete(name="") + 
  stat_summary(fun.data = "mean_sdl",geom='pointrange',fun.args = (mult=1),color='black',position = position_dodge(0.9)) + 
  ylab("number of unique reads") + 
  theme(text=element_text(size=18))


f.table <- as.data.frame(table(f.merge$sample))
f.table$group <- c(rep('scCT_brain',4),rep('scCT_cell_lines',2),rep('Kaya-Okur',3),rep('Grosselin scChIP',3))
f.table <- f.table[f.table$group != 'Kaya-Okur',]

# Hard-coding for revision, but no time
kaya_okur_per_replicate       <- data.frame(Var1=c('H3K27me3_K562_rep1','H3K27me3_K562_rep2','H3K27me3_H1','H3K4me2_rep1','H3K4me2_rep2'),
                                            Freq=c(908,479,486,489,393),
                                            group='Kaya-Okur')

f.table <- rbind(f.table,kaya_okur_per_replicate)
f.table$group <- factor(f.table$group,levels=unique(f.table$group)[c(1,2,4,3)])

f.table <- f.table[c(1:6,10:14,7:9),]
f.table$Var1 <- factor(f.table$Var1,levels=f.table$Var1)

p3 <- ggplot(data=f.table,aes(y=Freq,x=Var1,fill=group)) + 
  geom_bar(col='black',stat='identity') + theme_few() + 
  theme(axis.text.x = element_text(angle = 60,hjust=1)) + 
  xlab("") + 
  scale_fill_discrete(name="") + 
  ylab("") +
  theme(text=element_text(size=18))
p3

p1
p2
p3
ggsave(filename = paste0(params$out_prefix,'Figures/revision/scCT_QC_FriP.pdf'),plot = p1,width = 8,height = 6)
ggsave(filename = paste0(params$out_prefix,'Figures/revision/scCT_QC_nFragments.pdf'),plot = p2,width = 8,height = 6)
ggsave(filename = paste0(params$out_prefix,'Figures/revision/scCT_QC_nCells.pdf'),plot = p3,width = 8,height = 6)
```


################# Fingerprint plot

```{r}
library(reshape2)
library(ggplot2)
library(ggthemes)



samples = c('other_datasets/fingerprint_analysis/scCT_fingerprint.txt',
            'other_datasets/fingerprint_analysis/Kaya_okur_fingerprint.txt',
            'other_datasets/fingerprint_analysis/Grosselin_fingerprint.txt')


files.df      <- data.frame(samples=c("scCT","Kaya_okur","Grosselin"))
files.df$path <- paste0(params$out_prefix,samples)


to.plot <- c(c("K562_H3K27me3_iCell8.bam","H1_H3K27me3_iCell8.bam"),
             c("Astrocytes","Microglia","mOL","Neurons","Neurons","OEC","OPC","VLMC","/outs/possorted_bam.bam"),
             c("possorted_SRR7536860.bam"))


scCT_bulk <- "scCT_all_clusters.bam"
exp.names <- c('scCT','kaya-okur','Grosselin')


# Read files
df.ls <-lapply(files.df$path,function(x){
  NAMES       <- read.table(file=x,nrow=1,stringsAsFactors=F)
  D           <- as.data.frame(read.table(file=x,skip=2,stringsAsFactors=F))
  colnames(D) <- NAMES
  D[,grep(paste(to.plot,collapse="|"),colnames(D)),drop=FALSE]
})


# Make cumsum
df.ls.cumsum <- lapply(df.ls,function(x){
  x <- apply(x,2,sort)
  x <- apply(x,2,function(y){
    y <- cumsum(y)/sum(y)
    return(y)
  })
  x
})

scCT_names      <- c('scCT_bulk_N1','scCT_bulk_N2','scCT_bulk_N3','scCT_bulk_N4',colnames(df.ls.cumsum[[1]])[5:length(colnames(df.ls.cumsum[[1]]))])
kaya_okur_names <- c("K562_H3K27me3_iCell8","hESC_H1_H3K27me3_iCell8")
grosselin_names <- c("grosselin_H3K27me3")


colnames(df.ls.cumsum[[1]]) <-  scCT_names
colnames(df.ls.cumsum[[2]]) <-  kaya_okur_names
colnames(df.ls.cumsum[[3]]) <-  grosselin_names

# Melt
df.ls.cumsum.melt <- lapply(df.ls.cumsum,function(x){
  x <- melt(x)
  x$Var1 <- x$Var1/max(x$Var1)
  l <- dim(x)[1]
  s <- sort(sample(5:l-5, 1000))
  # Downsample the really big
  x <- rbind(head(x),x[s,],tail(x))
  x
})

# add names
df.ls.cumsum.melt <- lapply(1:length(df.ls.cumsum),function(x){
  df.ls.cumsum.melt[[x]]$group <- exp.names[x]
  df.ls.cumsum.melt[[x]]
})

df      <- do.call(rbind,df.ls.cumsum.melt)

# Fix names for plotting
df[grep('scCT_bulk_N',df$Var2),'group']   <- 'scCT_merged_replicate'
df[df$group=='scCT','group']               <- 'scCT_single_cluster'

#antibody column
df$antibody <- "H3K27me3"
df[grep("H3K4me2",df$Var2),'antibody'] <- "H3K4me2"


df$Var2 <- basename(as.character(df$Var2))
df$Var2 <- gsub(".bam","",df$Var2)


ggplot(data=df) + geom_line(aes(x=Var1,y=value,col=group,group=Var2)) + 
  coord_cartesian(xlim=c(0,max(df$Var1))) + 
  theme_few() + xlab("Genomic bins fraction") + ylab("Cummulative signal") + 
  theme(text=element_text(size=18),
        legend.title = element_blank())

ggsave(paste0(params$out_prefix,'Figures/revision/scCT_fingerprint.pdf'),width=8,height=4)

```


# Dispersion of coverage

```{r}
library(reshape2)
library(ggplot2)
library(ggthemes)



samples = c('other_datasets/fingerprint_analysis/scCT_fingerprint.txt',
            'other_datasets/fingerprint_analysis/Kaya_okur_fingerprint.txt',
            'other_datasets/fingerprint_analysis/Grosselin_fingerprint.txt')


files.df      <- data.frame(samples=c("scCT","Kaya_okur","Grosselin"))
files.df$path <- paste0(params$out_prefix,samples)


to.plot <- c(c("K562_H3K27me3_iCell8.bam","H1_H3K27me3_iCell8.bam"),
             c("Astrocytes","Microglia","mOL","Neurons","Neurons","OEC","OPC","VLMC","/outs/possorted_bam.bam"),
             c("possorted_SRR7536860.bam"))


scCT_bulk <- "scCT_all_clusters.bam"
exp.names <- c('scCT','kaya-okur','Grosselin')


# Read files
df.ls <-lapply(files.df$path,function(x){
  NAMES       <- read.table(file=x,nrow=1,stringsAsFactors=F)
  D           <- as.data.frame(read.table(file=x,skip=2,stringsAsFactors=F))
  colnames(D) <- NAMES
  D[,grep(paste(to.plot,collapse="|"),colnames(D)),drop=FALSE]
})

a <- hist(log10(df.ls[[2]][,2]))


```







################## PCA

# Cell lines
```{r,fig.width=6,fig.height=3}
library(reticulate)
library(ggplot2)
library(ggrepel)
library(ggthemes)


np <- import("numpy")
npz1 <- np$load(file = paste0(params$out_prefix,"other_datasets/PCA/PCA_table_cell_lines.npz"))

d <- npz1$f[['matrix']]
colnames(d) <- npz1$f[['labels']]

d <- d[order(matrixStats::rowVars(t(apply(d,1,function(x){x/sum(x)}))),decreasing = TRUE),]
d <- head(d,200)

d.pca <-prcomp(na.omit(d),scale=TRUE,center=TRUE)
plot.labels <- c("sc-3T3",
                 "sc-mES_A",
                 "sc-mES_B",
                 "sc-Oli-neu_A",
                 "sc-Oli-neu_B",
                 "bulk-Oli-neu",
                 "bulk-mESC_rep1",
                 "bulk-mESC_rep2",
                 "bulk-3T3_rep1",
                 "bulk-3T3_rep2")


pca.plot        <- as.data.frame(d.pca$rotation)
pca.plot$group  <- factor(c(rep("single_cell",5),rep("bulk",5)),levels=c('single_cell','bulk'))
pca.plot$sample <- plot.labels

p1 <- ggplot(data = pca.plot, aes(x = PC1,y = PC2,label = sample,shape=group)) + 
  geom_jitter(aes(color = group),size=5,stroke =1.5,width = 0.01,height=0.01) + 
  geom_text_repel(force=100,size=6,point.padding = 1) +
  theme_few() + 
  theme(legend.position = "None") + 
  coord_cartesian(xlim=c(-1.5,1.5),ylim=c(-0.8,0.5)) + 
  scale_shape_manual(values=c(3, 4))

p1 

d.plot <- cor(d)
colnames(d.plot) <- plot.labels
rownames(d.plot) <- plot.labels


b <- seq(from=-1,to=1,by=0.05)


ann.row <- data.frame(exp=unlist(lapply(strsplit(as.character(colnames(d.plot)),'-'),'[',1)))
rownames(ann.row) <- colnames(d.plot)

p2 <-pheatmap::pheatmap(d.plot,color = colorRampPalette(colors = c("blue","white","red"))(length(b)),breaks = b, treeheight_row = 0,
                        annotation_col =  ann.row,fontsize = 14)


print(gridExtra::grid.arrange(grobs=list(p1,p2[[4]]),nrow=1))

pdf(file = paste0(params$out_prefix,'Figures/revision/Cell_lines_PCA_correlation.pdf'),width=12,height=6)
par(mar=c(4,2,2,4))
print(gridExtra::grid.arrange(grobs=list(p1,p2[[4]]),nrow=1))
dev.off()
```




```{r, fig.width=10, fig.height=10}
library(matrixStats)
library(reshape2)

np <- import("numpy")
npz1 <- np$load(file = paste0(params$out_prefix,"other_datasets/correlation/H3K27me3_cell_lines_correlation_matrix.txt.gz"))

d <- npz1$f[['matrix']]
colnames(d) <- gsub(".bw","",npz1$f[['labels']])


d <- d[rowSums(d) > 0,]
colnames(d) <- gsub("_1_","_A_",colnames(d))
colnames(d) <- gsub("_2_","_B_",colnames(d))
#d.short <- d[sample(1:nrow(d),25000),]


d.short <- d[order(rowSums(d),decreasing = TRUE),]
d.short <- head(d.short,5000)

d.plot <- cor(d.short)
#d.plot[lower.tri(d.plot)] <- NA
d.plot <- melt(round(d.plot,digits = 2))

d.plot$Var1 <- factor(d.plot$Var1,levels=rev(levels(d.plot$Var1)))
d.plot$Var2 <- factor(d.plot$Var2,levels=levels(d.plot$Var2))

ggplot(data=d.plot,aes(x=Var1,y=Var2,fill=value)) + 
  geom_tile() + 
  scale_fill_gradient2(low = 'blue',high='red') + 
  theme_few() + 
  geom_text(aes(label=value),size=10) + 
  theme(axis.text.x = element_text(angle = 45,hjust=+1),
        text = element_text(size=30,face = 'bold')) + 
  xlab("") + 
  ylab("") + coord_flip()

ggsave(paste0(params$out_prefix,"Figures/revision/Cell_lines_correlation_heatmap.pdf"),width=16,height=16)
```

```{r, fig.width=8,fig.height=8,error=FALSE,warning=FALSE,message=FALSE}
library(GGally)


pdf(paste0(params$out_prefix,"Figures/revision/Cell_lines_correlation_scatter.pdf"),width=12,height=12)
ggpairs(as.data.frame(d.short),
                      #   upper = list(continuous=wrap("density",alpha=0.3),combo='box',col='black'),
                         upper = list('na'),
                         lower = list(continuous=wrap("points",alpha=0.1,size=0.1,col='black')),
                          diag = list('na')) + theme_few()
dev.off()

png(paste0(params$out_prefix,"Figures/revision/Cell_lines_correlation_scatter.png"),units = 'in',width = 16,height = 16,res = 150)
ggpairs(as.data.frame(d.short),
                      #   upper = list(continuous=wrap("density",alpha=0.3),combo='box',col='black'),
                         upper = list('na'),
                         lower = list(continuous=wrap("points",alpha=0.1,size=0.1,col='black')),
                          diag = list('na')) + theme_few()
dev.off()


```






```{r,fig.width=6,fig.height=3}
library(reticulate)
library(ggplot2)
library(ggthemes)
library(ggrepel)
library(pheatmap)

set.seed(2)
np <- import("numpy")
npz1 <- np$load(file = paste0(params$out_prefix,"other_datasets/PCA/PCA_table_brain.npz"))

d <- npz1$f[['matrix']]
colnames(d) <- npz1$f[['labels']]

#d <- d[,-grep("OEC",colnames(d))]
d <- d[order(matrixStats::rowVars(t(apply(d,1,function(x){x/sum(x)}))),decreasing = TRUE),]
d <- head(d,150)

d.pca <-prcomp(na.omit(d),scale=TRUE,center=TRUE)
plot.labels <- c("sc-Astrocytes",
                 "sc-Microglia",
                 "sc-Oligodendrocytes",
                 "sc-Neurons 1",
                 "sc-Neurons 2",
                 "sc-OEC",
                 "sc-OPC",
                 "sc-VLMC",
                 "bulk-Neurons",
                 "bulk-Microglia",
                 "bulk-OPC")
                 
                 
pca.plot        <- as.data.frame(d.pca$rotation)
pca.plot$group  <- c(rep("single_cell",8),rep("bulk",3))
pca.plot$sample <- plot.labels

p1 <- ggplot(data = pca.plot, aes(x = PC1,y = PC2,label = sample),size=3) + 
  geom_point(aes(color = group),size=3) + 
  geom_text_repel(force=5,point.padding=0.5,size=7) +
  coord_cartesian(xlim=c(-0.5,0.7),ylim=c(-0.25,0.5)) + 
  theme(legend.position = "None", 
        plot.background = element_blank(),
        panel.background = element_blank(),
        axis.line=element_line(),
        axis.ticks.length = unit(6,'points'))
p1  
#ggsave(plot = p,filename = "PCA_sc_bulk.png",width=5,height=5)

d.cor <- cor(d)
colnames(d.cor) <- plot.labels
rownames(d.cor) <- plot.labels


b <- seq(from=-1,to=1,by=0.05)
#p2 <- pheatmap::pheatmap(d.cor,color = colorRampPalette(colors = c("blue","white","red"))(length(b)),breaks = b,angle_col=45,treeheight_row = 0,fontsize = 16)
p2 <-pheatmap::pheatmap(d.cor,color = colorRampPalette(colors = c("blue","white","red"))(length(b)),breaks = b,treeheight_row = 0,fontsize = 14)

pdf(paste0(params$out_prefix,"Figures/revision/Brain_PCA.pdf"),width=12,height=6)
print(gridExtra::grid.arrange(grobs=list(p1,p2[[4]]),nrow=1))
dev.off()
```


```{r,fig.width=2,fig.height=3}
brain <- readRDS(paste0(params$out_prefix,"H3K27me3_cell_lines/clustering/01.clustering.Rds"))

idents.old <- unique(brain@active.ident)

idents <- idents.old
idents <- gsub("_1","_A",idents)
idents <- gsub("_2","_B",idents)
names(idents) <- idents.old

brain <- RenameIdents(object = brain,idents)
levels(brain@active.ident) <- c("mESC_A","mESC_B","3T3","Oli-neu_A","Oli-neu_B")

d.to.plot <- table(brain@active.ident,brain$sample)
d.to.plot <- apply(d.to.plot,2,function(x){x/sum(x)})
d.to.plot <- melt(d.to.plot)

ggplot(data=d.to.plot) + 
  geom_bar(aes(x=Var2,y=value,fill=Var1),stat='identity',col='black') + 
  theme_few() +
  ylab('fraction of total \n') + 
  xlab("") +
  scale_fill_manual(values = c('royalblue1','royalblue4','seagreen','maroon1','maroon4'),name='') + 
  scale_x_discrete(labels=c("Rep 1","Rep 2")) + 
  theme(legend.position = 'bottom',
        axis.text = element_text(size=20),
        axis.title.y = element_text(size=20),
        legend.text = element_text(size=16)) + 
  guides(fill=guide_legend(nrow=3,byrow = FALSE))
  

ggsave(paste0(params$out_prefix,"Figures/revision/Cell_lines_stacked_barplot.png"),width=4,height=6)
```





```{r}




# Colors definition - consistent accross the paper
CTcolors        <- scales::hue_pal()(9)


brain <- readRDS(paste0(params$out_prefix,'H3K27me3_cell_lines/clustering/01.clustering.Rds'))
DimPlot(brain)

brain$cell_type <- factor(brain$cell_type,levels=levels(brain$cell_type)[c(2,5,3,1,4)])
ggplot(brain@meta.data) + geom_boxplot(aes(y=logUMI,x=cell_type,fill=sample),outlier.shape = NA)  + theme_few() + scale_fill_discrete(name='',labels=c('Replicate 1','Replicate 2')) + theme(legend.position = 'top') + xlab('cluster')
paste0(params$out_prefix,"Figures/revision/Cell_lines_correlation_scatter.png")

brain$peak_region_ratio

ggplot(brain@meta.data,aes(y=peak_region_ratio,x=cell_type,fill=sample)) + 
  geom_violin(trim=FALSE)  + 
  theme_few() + 
  scale_fill_manual(name='',labels=c('Replicate 1','Replicate 2'),values = scales::hue_pal()(4)[c(2,4)]) + 
  theme(legend.position = 'top') + xlab('cluster') + 
  stat_summary(aes(x=cell_type,group=sample),fun.data = "mean_sdl",geom='pointrange',fun.args = (mult=1),color='black',position = position_dodge(0.9))



```


```{r}
get_percentage_cell_types <- function(x){
  sc.to.plot <- x@meta.data[,c('cell_type','sample'),]
  sc.to.plot <- aggregate(sc.to.plot$cell_type, by= list(sc.to.plot$sample,sc.to.plot$cell_type),length)
  
  sc.to.plot.percent.ls <- lapply(unique(sc.to.plot$Group.1),function(x){
    sc.to.plot.x <- sc.to.plot[sc.to.plot$Group.1 == x,]
    sc.to.plot.x$x <- sc.to.plot.x$x / sum(sc.to.plot.x$x)
    sc.to.plot.x
  })

sc.to.plot.percent <- do.call('rbind',sc.to.plot.percent.ls)
  
}

# order clusters
clusters_order <- c("Astrocytes","Vascular","mOL","Neurons_1","Neurons_2","Neurons_3","OEC","OPC","Microglia")

# Colors definition - consistent accross the paper
CTcolors        <- scales::hue_pal()(9)
names(CTcolors) <- clusters_order

# Load the data
scRNA <- readRDS(file = paste0(params$out_prefix,"Sox10_RNA/clustering/GFP/01.clustering.Rds"))
brain <- readRDS(file = paste0(params$out_prefix,"H3K27me3/clustering/01.clustering.Rds"))


# Rename the idents
scRNA <- RenameIdents(object = scRNA,"Oligodendrocytes"="mOL","VEC"="Vascular","Pericytes"="Vascular","VLMC"="Vascular")
scRNA$cell_type <- factor(as.character(scRNA@active.ident),levels=names(CTcolors))

brain.P25 <- brain[,brain$sample %in% c("H3K27me3_N1","H3K27me3_N2")]
brain.P25 <- RenameIdents(object = brain.P25,"VLMC"="Vascular")
brain.P25$cell_type <- factor(as.character(brain.P25@active.ident),levels=names(CTcolors))
brain.P25$cell_type <- droplevels(brain.P25$cell_type,exclude = c('Microglia','Neurons_1','Neurons_3','COP-NFOL')) # GFP+ sorting in N1,N2, so drop neurons, and microglia. COP-NFOLs dropped because scCT can't find them


# Process and plot

sc.to.plot <- get_percentage_cell_types(scRNA)
p1 <- ggplot(data=sc.to.plot) + geom_bar(aes(y=x/sum(x),x=Group.1,fill=Group.2),position='stack',stat='identity') + scale_color_manual(CTcolors)

sc.to.plot <- get_percentage_cell_types(brain.P25)
p2 <- ggplot(data=sc.to.plot) + geom_bar(aes(y=x/sum(x),x=Group.1,fill=Group.2),position='stack',stat='identity')+ scale_color_manual(CTcolors)

sc.to.plot.RNA <- cbind(get_percentage_cell_types(scRNA),exp='scRNA-seq')
sc.to.plot.CT  <- cbind(get_percentage_cell_types(brain.P25),exp='scCUT&Tag')

to.plot <- rbind(sc.to.plot.RNA,sc.to.plot.CT)
to.plot$Group.1 <- as.factor(to.plot$Group.1) 

ggplot(data = to.plot) + 
  geom_bar(aes(y=x,x=Group.1,fill=Group.2),position='stack',stat='identity',col='black') + 
  scale_fill_manual(name='',values = CTcolors) + 
  facet_grid(~exp,scales='free') + 
  theme_bw() + hh
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(angle = 45,hjust = 1),
        text = element_text(size=14)) + 
  ylab("fraction of total") + 
  xlab("")

ggsave(paste0(params$out_prefix,"Figures/revision/cell_types_scRNA_vs_scCT.pdf"),width=5,height=5)
```



