---
title: "R Notebook"
output:
  html_document:
    df_print: paged
params:
  config: '../../config/config.yaml'
  out_prefix: "~/mount/CT/git_test/results/" #'results/'
  antibody: 'H3K27me3'
  window: 5000
---

```{r libraries}
library(Seurat)
library(gridExtra)
library(dplyr)
library(Signac)
library(EnsDb.Mmusculus.v79)
library(ggplot2)
library(RColorBrewer)
library(rtracklayer)
library(yaml)
library(pheatmap)
library(GenomicFeatures)
library(tidyr)

getwd()
set.seed(100)
```


# Read config

```{r read_config}
source("../../scripts/R/func.R")
config <- yaml::read_yaml(params$config)

assay <- paste0("bins_",params$window)
```

# load annotations

```{r load_gene_annotations}
genebodyandpromoter.coords.flat <- load_ensembl_annot(config$general$Version)
```


```{r load_and_merge}
samples <- names(which(unlist(lapply(config$samples,function(x){x$Antibody == params$antibody}))))

brain.ls <- lapply(samples,function(x){
  # Read the seurat object
  brain                     <- readRDS(paste0(params$out_prefix,x,'/seurat/bin_',params$window,'/Seurat_object.Rds'))
  brain
})

brain.merged <- Reduce(function(x,y) merge(x,y), brain.ls)
DefaultAssay(brain.merged)       <- assay
```


```{r cluster_and_dimreduce_1}
# brain.merged <- brain.merged[which(rowSums(brain.merged[[assay]]@counts)>0),]
brain.merged <- RunTFIDF(brain.merged)
brain.merged <- FindTopFeatures(brain.merged, min.cutoff = 'q20')
  
brain.merged <- RunSVD(
  object = brain.merged,
  assay = assay,
  reduction.key = 'LSI_',
  reduction.name = 'lsi',
  n = 50
)
  
brain.merged <- RunUMAP(
  object = brain.merged,
  reduction = 'lsi',
  dims = 2:40
)

brain.merged <- FindNeighbors(
  object = brain.merged,
  reduction = 'lsi',
  dims = 2:50
)
brain.merged <- FindClusters(
  object = brain.merged,
  algorithm = 3,
  resolution = 0.4,
  verbose = FALSE
)

DimPlot(brain.merged,label=TRUE)
```

```{r vln_plot}
VlnPlot(brain.merged,'promoter_ratio',pt.size=0.1)

# Cluster 7 likely ATAC-seq signal - too high promoter ratio - remove from downstream analysis
# Also no hox genes methylation in cluster 7
```



```{r export_bw_for_inspection}
# Export bw to inspect cluster 7 merged profile

dir.create(paste0(params$out_prefix,params$antibody,"/clustering/bigwig_inspect/clusters_all/"),recursive = TRUE)

samples <- unlist(lapply(config$samples,function(x){x$Antibody}))
samples <- samples[samples == params$antibody]


fragments.path     <- paste0(params$out_prefix,params$antibody,'/fragments.tsv.gz')
fragments.grange   <- rtracklayer::import(con = fragments.path,format = 'bed')

chrom.sizes <- getChromInfoFromUCSC(genome = config$general$Version)
# chrom.sizes <- chrom.sizes[1:21,]


lapply(levels(brain.merged@active.ident),function(x){
  exportBW(object = brain.merged,
           cluster = x,
           fragments = fragments.grange,
           path =  paste0(params$out_prefix,params$antibody,"/clustering/bigwig_inspect/cluster_",x,".bw"),
           chrom.sizes = chrom.sizes)
})
```


```{r filter_UMI_promoters}
brain.merged <- brain.merged[,brain.merged$promoter_ratio < 0.6] # Filter out ATAC-seq like signal
brain.merged <- brain.merged[,brain.merged$logUMI >1.8]
```


```{r cluster_and_dimreduce_2}
# Re-do the dimreduction
brain.merged <- RunTFIDF(brain.merged)
brain.merged <- FindTopFeatures(brain.merged, min.cutoff = 'q20')
  
brain.merged <- RunSVD(
  object = brain.merged,
  assay = assay,
  reduction.key = 'LSI_',
  reduction.name = 'lsi',
  n = 50
)
  
brain.merged <- RunUMAP(
  object = brain.merged,
  reduction = 'lsi',
  dims = 2:40
)

brain.merged <- FindNeighbors(
  object = brain.merged,
  reduction = 'lsi',
  dims = 2:50
)
brain.merged <- FindClusters(
  object = brain.merged,
  algorithm = 3,
  resolution = 0.4,
  verbose = FALSE
)

DimPlot(brain.merged,label=TRUE)
```




```{r plots, fig.width=8,fig.height=4}

p1 <- DimPlot(brain.merged,label=TRUE)  + NoLegend()
p2 <- DimPlot(brain.merged,group.by='sample',label=TRUE) + theme(legend.position = 'bottom')

p3 <- DimPlot(brain.merged,group.by = 'Age')+ NoLegend() + theme(legend.position = 'bottom')
p4 <- DimPlot(brain.merged,group.by = 'GFP')+ NoLegend() + theme(legend.position = 'bottom') +
  scale_color_manual(values = RColorBrewer::brewer.pal(n = 9,name='Set1')[c(4,3)])

p.ls <- lapply(unique(brain.merged$sample),function(x){
  to_plot <- brain.merged[,brain.merged$sample == x]
  DimPlot(to_plot,label=TRUE,pt.size=0.3)+ NoLegend() + ggtitle(paste(x,unique(to_plot$Age), unique(to_plot$GFP))) + theme(plot.title = element_text(size=12)) + coord_cartesian(xlim=c(-10,15),ylim=c(-10,12))
})


p1 + p2
p3 + p4
do.call(grid.arrange,p.ls)
FeaturePlot(brain.merged,'logUMI',max.cutoff = 3) + scale_color_viridis_c()
```

```{r load_RNA_markers}
markers.Sox10 <- read.csv2(file=paste0(params$out_prefix,'Sox10_RNA/clustering/GFP/markers.csv'),row.names = 1)
markers       <- read.csv2(file=paste0(params$out_prefix,'Sten_RNA/clustering/sten_RNA_markers.csv'),row.names = 1)
```




```{r Plot_RNA_modules_projection,fig.width=6,fig.height=6}
assay='GA'

markers.Sox10$cluster <- paste0('Sox10_',markers.Sox10$cluster)
markers               <- rbind(markers,markers.Sox10)

brain.markers.ls <- lapply(unique(markers$cluster),function(x){
  marker.genes <- markers[markers$cluster==x & markers$p_val_adj < 0.05 & markers$avg_logFC > 0,"gene"]
  marker.genes <- head(marker.genes,200)
  marker.genes.agg.exp <- Matrix::colSums(brain.merged[[assay]]@data[rownames(brain.merged[[assay]]) %in% marker.genes,])
  marker.genes.agg.exp <- marker.genes.agg.exp / Matrix::colSums(brain.merged[[assay]]@counts)
  return(marker.genes.agg.exp)
})

names(brain.markers.ls)  <- gsub("-",".",unique(markers$cluster))
markers.agg.df           <- do.call(cbind,brain.markers.ls)
colnames(markers.agg.df) <- paste0("marker_",gsub(" ",".",colnames(markers.agg.df)))


brain.merged <- AddMetaData(brain.merged,metadata = markers.agg.df,col.name = colnames(markers.agg.df))

lapply(as.character(colnames(markers.agg.df)),function(x){
  FeaturePlot(brain.merged,x,max.cutoff = quantile(markers.agg.df[,x],0.9,na.rm=TRUE),min.cutoff = quantile(markers.agg.df[,x],0.05,na.rm=TRUE)) + scale_color_viridis_c() + NoLegend()+ coord_cartesian(xlim=c(-12,12),ylim=c(-12,12))
  #FeaturePlot(brain,x,max.cutoff = 0.08,min.cutoff = 0.02) + scale_color_viridis_c(limits=c(0.02,0.08)) 
  })

```

```{r dimplot}
DimPlot(brain.merged,label=TRUE)
```



```{r Rename_the_clusters,fig.width=10,fig.height=10}
# Remove small and weird clusters
brain.merged <- brain.merged[,!brain.merged@active.ident %in% c(9,10)]

# Rename the clusters
brain.merged <- RenameIdents(brain.merged, '0' = 'Astrocytes',
                                           '1' = 'mOL',
                                           '2' = 'OEC',
                                           '3' = 'Neurons_1',
                                           '4' = 'VLMC',
                                           '5' = 'OPC',
                                           '6' = 'mOL',
                                           '7' = 'Microglia',
                                           '8' = 'Neurons_3')

brain.merged$cell_type <- brain.merged@active.ident

p1 <- DimPlot(object = brain.merged, group.by='Age',label = TRUE,reduction = 'umap',pt.size=0.1) + NoLegend()
p2 <- DimPlot(object = brain.merged, label = TRUE,reduction = 'umap',pt.size=0.1) + NoLegend()
p3 <- DimPlot(object = brain.merged, group.by='GFP',label = TRUE,reduction = 'umap',pt.size=0.1) + NoLegend() + scale_color_manual(values = RColorBrewer::brewer.pal(n = 9,name='Set1')[c(4,3)])

p1
p2
p3
```




```{r find_markers_peaks}
DefaultAssay(brain.merged) <- 'peaks'

markers <- FindAllMarkers(brain.merged,min.pct = 0.01,logfc.threshold = 0.1,slot = 'data')    # Post-publication change min.pct=0.04 -> 0.01
markers$closest_gene <- ClosestFeature(regions = StringToGRanges(markers$gene ), annotation = genebodyandpromoter.coords.flat)$name

markers.top <- markers %>% group_by(cluster) %>% top_n(n = 50,wt = avg_logFC)
markers.negative <- markers %>% group_by(cluster) %>% top_n(n = 50,wt = -avg_logFC)

saveRDS(object = brain.merged,      file = paste0(params$out_prefix,params$antibody,"/clustering/01.clustering.Rds"))
write.csv2(  x = markers.top,       file = paste0(params$out_prefix,params$antibody,"/clustering/peaks_markers_top.csv"))

```


# Export marker bed file 

```{r export_bed_marker_peaks}
dir.create(paste0(params$out_prefix,params$antibody,"/clustering/markers_bed/"),recursive = TRUE)
clusters_order = levels(brain.merged@active.ident)

# Get top markers
top.markers <- markers %>% group_by(cluster) %>% dplyr::filter(avg_logFC > 0) %>% dplyr::filter(p_val_adj < 0.05) %>% top_n(200, wt=avg_logFC) %>% drop_na()


top.markers         <- as.data.frame(top.markers)
top.markers         <- top.markers[order(top.markers$cluster),]


lapply(unique(top.markers$cluster),function(x){
  top.markers.x <- top.markers %>% dplyr::filter(cluster == x) %>% drop_na()
  top.markers.x <- sort(StringToGRanges(top.markers.x$gene,sep = c('-','-')))
  rtracklayer::export.bed(object = top.markers.x,con = paste0(params$out_prefix,params$antibody,"/clustering/markers_bed/markers_cluster_",x,".bed"))
})
```



```{r export_bw}
dir.create(paste0(params$out_prefix,params$antibody,"/clustering/bigwig/clusters_all/"),recursive = TRUE)

samples <- unlist(lapply(config$samples,function(x){x$Antibody}))
samples <- samples[samples == params$antibody]


fragments.path     <- paste0(params$out_prefix,params$antibody,'/fragments.tsv.gz')
fragments.grange   <- rtracklayer::import(con = fragments.path,format = 'bed')

chrom.sizes <- getChromInfoFromUCSC(genome = config$general$Version)
# chrom.sizes <- chrom.sizes[1:21,]


lapply(levels(brain.merged@active.ident),function(x){
  exportBW(object = brain.merged,
           cluster = x,
           fragments = fragments.grange,
           path =  paste0(params$out_prefix,params$antibody,"/clustering/bigwig/cluster_",x,".bw"),
           chrom.sizes = chrom.sizes)
})


fragments.x <- fragments.grange[fragments.grange$name %in% colnames(brain.merged)]
coverage.x  <- coverage(fragments.x) / (length(fragments.x)/1e6)
rtracklayer::export.bw(object = coverage.x,con = paste0(params$out_prefix,params$antibody,"/clustering/bigwig/clusters_all/clusters_all.bw"))

```


```{r find_markers}
DefaultAssay(brain.merged) <- paste0('bins_',params$window)
brain.merged<- NormalizeData(brain.merged,normalization.method = 'LogNormalize',scale.factor=10000)
markers <- FindAllMarkers(brain.merged,min.pct = 0.04,logfc.threshold = 0.1,slot = 'data')

markers$closest_gene <- ClosestFeature(regions = StringToGRanges(markers$gene ), annotation = genebodyandpromoter.coords.flat)$name
#View(markers)

markers.top <- markers %>% group_by(cluster) %>% top_n(n = 50,wt = avg_logFC)
markers.negative <- markers %>% group_by(cluster) %>% top_n(n = 50,wt = -avg_logFC)

#saveRDS(object = brain.merged,      file = paste0(params$out_prefix,params$antibody,"/clustering/01.clustering.Rds"))
write.csv2(  x = markers,           file = paste0(params$out_prefix,params$antibody,"/clustering/markers.csv"))
write.csv2(  x = markers.top,       file = paste0(params$out_prefix,params$antibody,"/clustering/markers_top.csv"))
write.csv2(  x = markers.negative,  file = paste0(params$out_prefix,params$antibody,"/clustering/markers_negative.csv"))



```

```{r do_heatmap}
ncells        <- dim(brain.merged)[2]
cells_to_plot <- colnames(brain.merged)[sample(1:ncells,ncells * 0.4)]
cells_to_plot <- tibble::rownames_to_column(brain.merged[,cells_to_plot]@meta.data) %>% arrange(seurat_clusters,sample) %>% pull(rowname)
clusters_order <- levels(brain.merged@active.ident)

# Get top markers
top.markers <- markers %>% group_by(cluster) %>% dplyr::filter(avg_logFC > 0) %>% top_n(100, wt=avg_logFC)
top.markers <- top.markers[which(!top.markers$gene %in% names(table(top.markers$gene)[table(top.markers$gene) > 1])),] # remove non-unique

# Reorder markers
top.markers$cluster <- factor(as.character(top.markers$cluster),levels=clusters_order)
top.markers         <- top.markers[order(top.markers$cluster),]

# Reorder cells
Idents(brain.merged) <- factor(as.character(Idents(brain.merged)),levels=clusters_order)

# Create list of cluster colors
clusterColors <- rep(list(scales::hue_pal()(length(levels(brain.merged@active.ident)))),2)
names(clusterColors) <- c("markers","clusters")
  
names(clusterColors[[1]]) <- levels(brain.merged@active.ident)
names(clusterColors[[2]]) <- levels(brain.merged@active.ident)


png(filename = paste0(params$out_prefix,params$antibody,"/clustering/bins_heatmap.png"),width=1000,height=1000)
DoHeatmapMB(brain.merged,
            markers = top.markers,
            assay = 'bins_5000',
            clusterColors = clusterColors,
            slot = 'data')
dev.off()

DoHeatmapMB(brain.merged,
            markers = top.markers,
            clusterColors = clusterColors,
            slot = 'data',
            assay = 'bins_5000')


```



```{r}
cluster.annotations           <- data.frame(cluster=brain.merged@active.ident,barcode=names(brain.merged@active.ident))
colnames(cluster.annotations) <- paste0("#",colnames(cluster.annotations))

write.csv(x = cluster.annotations, file=paste0(params$out_prefix,params$antibody,'/clustering/bam_per_cluster/cluster_barcode_table.csv'),row.names = FALSE,quote = FALSE,)

```











