---
title: "R Notebook"
output:
  html_document:
    df_print: paged
params:
  config: '../../config/config.yaml'
  out_prefix: "~/mount/CT/git_test/results/" #'results/'
  antibody: 'H3K27me3_cell_lines'
  window: 5000
---
  
```{r libraries}
library(Seurat,quietly = TRUE,warn.conflicts = FALSE)
library(gridExtra,quietly = TRUE,warn.conflicts = FALSE)
library(dplyr,quietly = TRUE,warn.conflicts = FALSE)
library(Signac,quietly = TRUE,warn.conflicts = FALSE)
library(EnsDb.Mmusculus.v79,quietly = TRUE,warn.conflicts = FALSE)
library(ggplot2,quietly = TRUE,warn.conflicts = FALSE)
library(rtracklayer,quietly = TRUE,warn.conflicts = FALSE)
library(yaml,quietly = TRUE,warn.conflicts = FALSE)
library(pheatmap,quietly = TRUE,warn.conflicts = FALSE)
set.seed(100)

window   = 5000
assay    = paste0('bins_',window)
antibody = params$antibody
```


# Read config

```{r read_config}
config <- yaml::read_yaml(params$config)
```

# load annotations

```{r load_gene_annotations}
gene.coords <- ensembldb::genes(EnsDb.Mmusculus.v79, filter = ~ gene_biotype == "protein_coding")
lncRNA.coords <- ensembldb::genes(EnsDb.Mmusculus.v79, filter = ~ gene_biotype == "lincRNA")
gene.coords <- c(gene.coords,lncRNA.coords)

seqlevelsStyle(gene.coords) <- 'UCSC'
genebody.coords <- keepStandardChromosomes(gene.coords, pruning.mode = 'coarse')

# Flatten the overlapping genes and extend by 2kb upstream of promoters
genebody.coords.flat <- GenomicRanges::reduce(x = genebody.coords)
genebodyandpromoter.coords.flat <- Signac::Extend(genebody.coords.flat,upstream = 2000)

# Retrieve gene names from the original annotation (lost because of flatenning)
genebodyandpromoter.coords.flat$name<- gene.coords[nearest(genebodyandpromoter.coords.flat,genebody.coords)]$gene_name
```

```{r load_and_merge}
samples <- names(which(unlist(lapply(config$samples,function(x){x$Antibody == params$antibody}))))



brain.ls <- lapply(samples,function(x){
  # Read the seurat object
  brain                     <- readRDS(paste0(params$out_prefix,x,'/cell_picking/',window,'/Seurat_object.Rds'))
  DefaultAssay(brain)       <- assay
  brain
})

brain.merged <- Reduce(function(x,y) merge(x,y), brain.ls)
```

```{r filter_UMI_promoters}
brain.merged <- brain.merged[,brain.merged$logUMI > 2.3]
```

# Add promoter activity assay

```{r promoter_activity_assay}
# Fetch promoter coordinates
promoter.coords <- promoters(EnsDb.Mmusculus.v79)
promoter.coords <- promoter.coords[promoter.coords$tx_biotype == "protein_coding" | promoter.coords$tx_biotype== "lincRNA"]
seqlevelsStyle(promoter.coords) <- 'UCSC'

fragments.path <- paste0(params$out_prefix,'/merged/',params$antibody,'/fragments.tsv.gz')
promoter.matrix <- FeatureMatrix(fragments = fragments.path,features = promoter.coords,cells = colnames(brain.merged))

promoters.key <- promoter.coords$tx_name
names(promoters.key) <- GRangesToString(promoter.coords)
rownames(promoter.matrix) <- promoters.key[rownames(promoter.matrix)]
promoter.matrix <- promoter.matrix[rownames(promoter.matrix) != "",]
brain.merged[['PA']] <- CreateAssayObject(counts = promoter.matrix)
```


```{r cluster_and_dimreduce}
# Re-do the dimreduction
brain.merged <- RunTFIDF(brain.merged)
brain.merged <- FindTopFeatures(brain.merged, min.cutoff = 'q20')
  
brain.merged <- RunSVD(
  object = brain.merged,
  assay = assay,
  reduction.key = 'LSI_',
  reduction.name = 'lsi',
  n = 50
)
  
brain.merged <- RunUMAP(
  object = brain.merged,
  reduction = 'lsi',
  dims = 2:30
)

brain.merged <- FindNeighbors(
  object = brain.merged,
  reduction = 'lsi',
  dims = 2:30
)
brain.merged <- FindClusters(
  object = brain.merged,
  algorithm = 3,
  resolution = 0.2,
  verbose = FALSE
)

DimPlot(brain.merged,label=TRUE)
```


```{r peaks_dimreduce}
peaks <- rtracklayer::import(paste0(params$out_prefix,'/merged/',params$antibody,'/macs/broad/',params$antibody,'_peaks.broadPeak'))
peaks <- subsetByOverlaps(x = peaks,ranges =  blacklist_mm10,invert = TRUE)

fragments <- paste0(params$out_prefix,unique(brain.merged$orig.ident),'/outs/fragments.tsv.gz')
names(fragments) <- unique(brain.merged$orig.ident)


peak.matrix.ls <- lapply(names(fragments),function(x){
  FeatureMatrix(fragments = fragments[x],features = peaks,cells = colnames(brain.merged[,brain.merged$orig.ident == x]))
})

rows.summary <- table(unlist(lapply(peak.matrix.ls,rownames))) 
common.rows  <- names(rows.summary[rows.summary == length(peak.matrix.ls)])

peak.matrix.ls<-  lapply(peak.matrix.ls,function(x){
  x[common.rows,]
})

peak.matrix.merged <- do.call(cbind,peak.matrix.ls)
brain.merged[['peaks']] <- CreateAssayObject(counts = peak.matrix.merged)


assay <- 'peaks'
DefaultAssay(brain.merged) <- assay
brain.merged <- RunTFIDF(brain.merged)
brain.merged <- FindTopFeatures(brain.merged, min.cutoff = 'q0')
  
brain.merged.peaks <- RunSVD(
  object = brain.merged,
  assay = assay,
  reduction.key = 'LSI_',
  reduction.name = 'lsi'
)
  
brain.merged.peaks <- RunUMAP(
  object = brain.merged.peaks,
  reduction = 'lsi',
  dims = 2:30
)

  
DimPlot(brain.merged.peaks,label=TRUE)
DimPlot(brain.merged.peaks,group.by='sample') + theme(legend.position = 'bottom')
```


```{r plots, fig.width=8,fig.height=4}

p1 <- DimPlot(brain.merged,label=TRUE)  + NoLegend()
p2 <- DimPlot(brain.merged,group.by='sample',label=TRUE) + theme(legend.position = 'bottom')

p3 <- DimPlot(brain.merged,group.by = 'Age')+ NoLegend() + theme(legend.position = 'bottom')
p4 <- DimPlot(brain.merged,group.by = 'GFP')+ NoLegend() + theme(legend.position = 'bottom') +
  scale_color_manual(values = RColorBrewer::brewer.pal(n = 9,name='Set1')[c(4,3)])

p.ls <- lapply(unique(brain.merged$sample),function(x){
  to_plot <- brain.merged[,brain.merged$sample == x]
  DimPlot(to_plot,label=TRUE,pt.size=0.3)+ NoLegend() + ggtitle(paste(x,unique(to_plot$Age), unique(to_plot$GFP))) + theme(plot.title = element_text(size=12)) + coord_cartesian(xlim=c(-10,15),ylim=c(-10,12))
})


p1 + p2
p3 + p4
do.call(grid.arrange,p.ls)
FeaturePlot(brain.merged,'logUMI',max.cutoff = 3) + scale_color_viridis_c()
```



```{r Rename_the_clusters,fig.width=6,fig.height=6}
# Remove small and weird clusters

#brain.merged <- brain.merged[,!brain.merged@active.ident %in% c(9,10)]

# Rename the clusters
brain.merged <- RenameIdents(brain.merged, '0' = 'Oli-neu_1',
                                           '1' = 'mESC_1',
                                           '2' = '3T3',
                                           '3' = 'Oli-neu_2',
                                           '4' = 'mESC_2')

brain.merged$cell_type <- brain.merged@active.ident

p1 <- DimPlot(object = brain.merged, group.by='Age',label = TRUE,reduction = 'umap',pt.size=0.1) + NoLegend()
p2 <- DimPlot(object = brain.merged, label = TRUE,reduction = 'umap',pt.size=0.1) + NoLegend()
p3 <- DimPlot(object = brain.merged, group.by='GFP',label = TRUE,reduction = 'umap',pt.size=0.1) + NoLegend() + scale_color_manual(values = RColorBrewer::brewer.pal(n = 9,name='Set1')[c(4,3)])

p1
p2
p3
```



```{r find_markers_peaks}
DefaultAssay(brain.merged) <- 'peaks'
brain.merged<- NormalizeData(brain.merged,normalization.method = 'LogNormalize',scale.factor=10000)
markers <- FindAllMarkers(brain.merged,min.pct = 0.04,logfc.threshold = 0.1,slot = 'data')

markers$closest_gene <- ClosestFeature(StringToGRanges(markers$gene ), genebodyandpromoter.coords.flat)$name
#View(markers)

markers.top <- markers %>% group_by(cluster) %>% top_n(n = 50,wt = avg_logFC)
markers.negative <- markers %>% group_by(cluster) %>% top_n(n = 50,wt = -avg_logFC)

saveRDS(object = brain.merged,      file = paste0(params$out_prefix,params$antibody,"/clustering/01.clustering.Rds"))
write.csv2(  x = markers.top,       file = paste0(params$out_prefix,params$antibody,"/clustering/peaks_markers_top.csv"))

```


```{r export_bw}
samples <- unlist(lapply(config$samples,function(x){x$Antibody}))
samples <- samples[samples == params$antibody]

# fragments.ls <- lapply(names(samples),function(x){
#   fragments <- paste0(params$out_prefix,x,'/outs/fragments.tsv.gz')
#   fragments <- rtracklayer::import(fragments,format = 'bed')
#   fragments
# })
# fragments <- Reduce(function(x,y) c(x,y), fragments.ls)
# fragments <- sort(fragments)

fragments <- rtracklayer::import(con = fragments.path,format = 'bed')

chrom.sizes <- read.table(url('http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/mm10.chrom.sizes'),sep="\t",stringsAsFactors = FALSE)
chrom.sizes <- chrom.sizes[1:21,]

exportBW <- function(object,cluster,fragments){
  if(class(object) == "Seurat"){
    cells <- rownames(object@meta.data[object@active.ident == cluster,])
  }
  
  new_read <- GRanges(seqnames = chrom.sizes[,1], 
        ranges =IRanges(start = as.numeric(chrom.sizes[,2]),
                        width=1),
        name = rep("in_silico_extra_read",dim(chrom.sizes)[1]),
        score = rep(0,dim(chrom.sizes)[1])
        )
  
  fragments.x <- fragments$name %in% cells
  fragments.x <- fragments[fragments.x]
  fragments.x <- c(fragments.x,new_read)
  
  
  coverage.x <- coverage(fragments.x)
  
  # Option A - normalize by number of reads per sample
  coverage.x <- coverage.x/(length(fragments.x)/1e6)
  
  # Option B - normalize by mean signal (~ enrichment of mean signal)
  # coverage.x <- coverage.x / mean(unlist(coverage.x))
  
  rtracklayer::export.bw(object = coverage.x,paste0(params$out_prefix,params$antibody,"/clustering/bigwig/cluster_",cluster,".bw"))
}

lapply(levels(brain.merged@active.ident),function(x){
  exportBW(brain.merged,x,fragments)
})

dir.create(paste0(params$out_prefix,params$antibody,"/clustering/bigwig/clusters_all/"))
fragments.x <- fragments[fragments$name %in% colnames(brain.merged)]
coverage.x  <- coverage(fragments.x) / length(fragments.x)
rtracklayer::export.bw(object = coverage.x,con = paste0(params$out_prefix,params$antibody,"/clustering/bigwig/clusters_all/clusters_all.bw"))

```

# Export marker bed file 

```{r export_bed}
clusters_order = levels(brain.merged@active.ident)

# Get top markers
top.markers <- markers %>% group_by(cluster) %>% dplyr::filter(avg_logFC > 0) %>% dplyr::filter(p_val_adj < 0.05) %>% top_n(200, wt=avg_logFC)

# Reorder markers
top.markers$cluster <- factor(as.character(top.markers$cluster),levels=clusters_order[1:sum(table(top.markers$cluster) > 0)])

top.markers         <- as.data.frame(top.markers)
top.markers         <- top.markers[order(top.markers$cluster),]


lapply(levels(top.markers$cluster),function(x){
  top.markers.x <- top.markers[top.markers$cluster == x,]
  top.markers.x <- sort(StringToGRanges(top.markers.x$gene))
  rtracklayer::export.bed(object = top.markers.x,con = paste0(params$out_prefix,params$antibody,"/clustering/markers_bed/markers_cluster_",x,".bed"))
})
```




```{r find_markers}
DefaultAssay(brain.merged) <- paste0('bins_',params$window)
brain.merged<- NormalizeData(brain.merged,normalization.method = 'LogNormalize',scale.factor=10000)
markers <- FindAllMarkers(brain.merged,min.pct = 0.04,logfc.threshold = 0.1,slot = 'data')

markers$closest_gene <- ClosestFeature(StringToGRanges(markers$gene ), genebodyandpromoter.coords.flat)$name
#View(markers)

markers.top <- markers %>% group_by(cluster) %>% top_n(n = 50,wt = avg_logFC)
markers.negative <- markers %>% group_by(cluster) %>% top_n(n = 50,wt = -avg_logFC)

#saveRDS(object = brain.merged,      file = paste0(params$out_prefix,params$antibody,"/clustering/01.clustering.Rds"))
write.csv2(  x = markers,           file = paste0(params$out_prefix,params$antibody,"/clustering/markers.csv"))
write.csv2(  x = markers.top,       file = paste0(params$out_prefix,params$antibody,"/clustering/markers_top.csv"))
write.csv2(  x = markers.negative,  file = paste0(params$out_prefix,params$antibody,"/clustering/markers_negative.csv"))



```

```{r do_heatmap}

DoHeatmapMB <- function(object, markers,fraction = 0.2,slot = "data",assay, clusterColors){
  
  object <- object[,sample(colnames(object),length(colnames(object)) * fraction)]
  
  row_annotation.df           <- data.frame(markers=markers$cluster)
  rownames(row_annotation.df) <- markers$gene
  col_annotation.df           <- data.frame(clusters=Idents(object))
  col_annotation.df$sample    <- object$sample
  col_annotation.df           <- col_annotation.df[order(col_annotation.df$clusters,col_annotation.df$sample),]
  
  heatmap.mat <- GetAssayData(object = object, slot,assay = assay)
  heatmap.mat <- heatmap.mat[rownames(row_annotation.df),rownames(col_annotation.df)]
  
  print(dim(heatmap.mat))
  
  pheatmap::pheatmap(heatmap.mat,
                     cluster_cols = FALSE, 
                     cluster_rows = FALSE,
                     annotation_row = row_annotation.df,
                     annotation_col = col_annotation.df,
                     color = viridis::viridis(256),
                     show_rownames = FALSE,
                     show_colnames = FALSE,
                     labels_col = "",labels_row = "",
                     annotation_colors = clusterColors)
}

ncells        <- dim(brain.merged)[2]
cells_to_plot <- colnames(brain.merged)[sample(1:ncells,ncells * 0.4)]
cells_to_plot <- tibble::rownames_to_column(brain.merged[,cells_to_plot]@meta.data) %>% arrange(seurat_clusters,sample) %>% pull(rowname)
clusters_order <- levels(brain.merged@active.ident)

# Get top markers
top.markers <- markers %>% group_by(cluster) %>% dplyr::filter(avg_logFC > 0) %>% top_n(100, wt=avg_logFC)
top.markers <- top.markers[which(!top.markers$gene %in% names(table(top.markers$gene)[table(top.markers$gene) > 1])),] # remove non-unique

# Reorder markers
top.markers$cluster <- factor(as.character(top.markers$cluster),levels=clusters_order)
top.markers         <- top.markers[order(top.markers$cluster),]

# Reorder cells
Idents(brain.merged) <- factor(as.character(Idents(brain.merged)),levels=clusters_order)

# Create list of cluster colors
clusterColors <- rep(list(scales::hue_pal()(length(levels(brain.merged@active.ident)))),2)
names(clusterColors) <- c("markers","clusters")
  
names(clusterColors[[1]]) <- levels(brain.merged@active.ident)
names(clusterColors[[2]]) <- levels(brain.merged@active.ident)


png(filename = paste0(params$out_prefix,params$antibody,"/clustering/bins_heatmap.png"),width=1000,height=1000)
DoHeatmapMB(brain.merged,
            markers = top.markers,
            assay = 'bins_5000',
            clusterColors = clusterColors,
            slot = 'data')
dev.off()

DoHeatmapMB(brain.merged,
            markers = top.markers,
            clusterColors = clusterColors,
            slot = 'data',
            assay = 'bins_5000')


```



```{r}
cluster.annotations           <- data.frame(cluster=brain.merged@active.ident,barcode=names(brain.merged@active.ident))
colnames(cluster.annotations) <- paste0("#",colnames(cluster.annotations))

write.csv(x = cluster.annotations, file=paste0(params$out_prefix,params$antibody,'/clustering/bam_per_cluster/cluster_barcode_table.csv'),row.names = FALSE,quote = FALSE,)

```











